src/i2c/src/i2c.rs

```diff
@@ -0,0 +1,516 @@
+// successful build for other non-musl libraries.
+#[cfg(target_env = "musl")]
+type IoctlRequest = c_int;
+#[cfg(not(target_env = "musl"))]
+type IoctlRequest = c_ulong;
+
+/// Standard I2C definitions
+/// I2C IOCTL commands
+const I2C_SLAVE: IoctlRequest = 0x0703; // Use this slave address
+const I2C_FUNCS: IoctlRequest = 0x0705; // Get the adapter functionality mask
+const I2C_RDWR: IoctlRequest = 0x0707; // Combined R/W transfer (one STOP only)
+const I2C_SMBUS: IoctlRequest = 0x0720; // SMBus transfer
+
+/// i2c/SMBUS Functions
+const I2C_FUNC_I2C: u64 = 0x00000001;
+const I2C_FUNC_SMBUS_READ_BYTE: u64 = 0x00020000;
+const I2C_FUNC_SMBUS_WRITE_BYTE: u64 = 0x00040000;
+const I2C_FUNC_SMBUS_READ_BYTE_DATA: u64 = 0x00080000;
+const I2C_FUNC_SMBUS_WRITE_BYTE_DATA: u64 = 0x00100000;
+const I2C_FUNC_SMBUS_READ_WORD_DATA: u64 = 0x00200000;
+const I2C_FUNC_SMBUS_WRITE_WORD_DATA: u64 = 0x00400000;
+
+const I2C_FUNC_SMBUS_BYTE: u64 = I2C_FUNC_SMBUS_READ_BYTE | I2C_FUNC_SMBUS_WRITE_BYTE;
+const I2C_FUNC_SMBUS_BYTE_DATA: u64 =
+    I2C_FUNC_SMBUS_READ_BYTE_DATA | I2C_FUNC_SMBUS_WRITE_BYTE_DATA;
+const I2C_FUNC_SMBUS_WORD_DATA: u64 =
+    I2C_FUNC_SMBUS_READ_WORD_DATA | I2C_FUNC_SMBUS_WRITE_WORD_DATA;
+
+/// SMBUS definitions
+/// SMBUS read or write markers
+const I2C_SMBUS_WRITE: u8 = 0;
+const I2C_SMBUS_READ: u8 = 1;
+
+/// SMBus transaction types (size parameter in the above functions)
+const I2C_SMBUS_QUICK: u32 = 0;
+const I2C_SMBUS_BYTE: u32 = 1;
+const I2C_SMBUS_BYTE_DATA: u32 = 2;
+const I2C_SMBUS_WORD_DATA: u32 = 3;
+
+/// As specified in SMBus standard
+const I2C_SMBUS_BLOCK_MAX: usize = 32;
+
+union I2cSmbusData {
+    byte: u8,
+    word: u16,
+    block: [u8; I2C_SMBUS_BLOCK_MAX + 2],
+}
+
+#[repr(C)]
+struct I2cSmbusIoctlData {
+    read_write: u8,
+    command: u8,
+    size: u32,
+    data: *mut I2cSmbusData,
+}
+
+/// I2C definitions
+pub const I2C_M_RD: u16 = 0x0001; // read data, from slave to master
+
+#[derive(Copy, Clone)]
+#[repr(C)]
+pub struct I2cMsg {
+    pub addr: u16,
+    pub flags: u16,
+    pub len: u16,
+    pub buf: *mut u8,
+}
+
+#[repr(C)]
+struct I2cRdwrIoctlData {
+    msgs: *mut I2cMsg,
+    nmsgs: u32,
+}
+
+#[repr(C)]
+pub struct I2cReq {
+    pub addr: u16,
+    pub flags: u16,
+    pub len: u16,
+    pub buf: Vec<u8>,
+}
+
+// Local I2C definitions
+const MAX_I2C_VDEV: usize = 1 << 7;
+const I2C_INVALID_ADAPTER: u32 = 0xFFFFFFFF;
+
+#[derive(Debug, Copy, Clone)]
+struct I2cAdapter {
+    fd: i32,
+    bus: u32,
+    smbus: bool,
+}
+
+pub struct I2cBackend {
+    adapters: Vec<I2cAdapter>,
+    adapter_map: [u32; MAX_I2C_VDEV],
+}
+
+fn vi2c_set_client_addr(fd: i32, addr: usize) -> errno::Result<()> {
+    let ret = unsafe { ioctl(fd, I2C_SLAVE, addr as c_ulong) };
+    if ret == -1 {
+        error!("Failed to set client addr to {:x}", addr);
+        errno::errno_result()
+    } else {
+        Ok(())
+    }
+}
+
+fn i2c_xfer(adapter: &I2cAdapter, reqs: &mut [I2cReq]) -> errno::Result<()> {
```

Comment:
Can this be a method implemented in `I2cAdapter` instead of passing `I2cAdapter as a parameter`? Same for the other functions that take a reference to `I2cAdapter`.

Reply:
> Can this be a method implemented in `I2cAdapter` instead of passing `I2cAdapter as a parameter`? Same for the other functions that take a reference to `I2cAdapter`.

Done